standards/application/waku-api.md

```diff
@@ -0,0 +1,327 @@
+
+The API defined in this document is an opinionated-by-purpose method to use the more agnostic [WAKU2]() protocols.
+
+## Syntax
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, 
+“RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## API design
+
+### IDL
+
+A custom Interface Definition Language (IDL) in YAML is used to define the Waku API.
+Existing IDL Such as OpenAPI, AsyncAPI or WIT do not exactly fit the requirements for this API.
+Hence, instead of having the reader learn a new IDL, we propose to use a simple IDL with self-describing syntax.
+
+An alternative would be to choose a programming language. However, such choice may express unintended opinions on the API.
+
+### Primitive types and general guidelines
+
+- No `default` means that the value is mandatory, meaning a `default` value implies an optional parameter.
+- Primitive types are `string`, `int`, `bool`, `enum` and `uint`
+- Complex pre-defined types are:
+  - `struct`: object and other nested types.
+  - `array`: iterable object containing values of all the same type.
+  - `result`: an enum type that either contains a value or void (success), or an error (failure); The error is left to the implementor.
+  - `error`: Left to the implementor on whether `error` types are `string` or `object` in the given language.
+- Usage of `result` is RECOMMENDED, usage of exceptions is NOT RECOMMENDED, no matter the language.
+
+### Language mappings
+
+How the API definition should be translated to specific languages.
+
+```yaml
+language_mappings:
+  typescript:
+    naming_convention:
+      - functions: "camelCase"
+      - variables: "camelCase"
+      - types: "PascalCase"
+  nim:
+    naming_convention:
+      - functions: "camelCase"
+      - variables: "camelCase"
+      - types: "PascalCase"
+```
+
+### Application
+
+This API is designed for generic use and ease across all programming languages, for `edge` and `relay` type nodes.
+
+## The Waku API
+
+```yaml
+api_version: "0.0.1"
+library_name: "waku"
+description: "Waku: a private and censorship-resistant message routing library."
+```
+
+### Initialise Waku node
+
+#### Type definitions
+
+```yaml
+types:
+  WakuNode:
+    type: struct
+    description: "A Waku node instance."
+
+  NodeConfig:
+    type: struct
+    fields:
+      mode:
+        type: string
+        constraints: [ "edge", "relay" ]
+        default: "_platform dependent_"
+        description: "The mode of operation of the Waku node. Core protocols used by the node are inferred from this mode."
+      waku_config:
+        type: WakuConfig
+        default: TheWakuNetworkPreset
+      message_confirmation:
+        type: array<string>
+        constraints: [ "store", "filter" ]
+        # Until further dogfooding, assuming default false, usage of SDS should be preferred
```

Comment:
I don't understand why we mention `SDS` here as it is in `dogfooding` / `experimental` stage too